SimPEG/directives.py

```diff
@@ -1173,3 +1406,264 @@ def endIter(self):
+    '''
+    header = None
+    save_txt = True
+    betas = None
+    phi_d = None
+    phi_m = None
+    phi_c = None
+    phi = None
+
+    def initialize(self):
+        if self.save_txt is True:
+            print(
+                "SimPEG.SaveOutputEveryIteration will save your inversion "
+                "progress as: '###-{0!s}.txt'".format(self.fileName)
+            )
+            f = open(self.fileName+'.txt', 'w')
+            self.header = "  #          betas            lambda         joint_phi_d                joint_phi_m            phi_c       iterCG     phi    \n"
+            f.write(self.header)
+            f.close()
+
+        # Create a list of each
+        self.betas = []
+        self.lambd = []
+        self.phi_d = []
+        self.phi_m = []
+        self.phi = []
+        self.phi_c = []
+
+    def endIter(self):
+
+        self.betas.append(["{:.2e}".format(elem) for elem in self.invProb.betas])
+        self.phi_d.append(["{:.3e}".format(elem) for elem in self.invProb.phi_d_joint])
+        self.phi_m.append(["{:.3e}".format(elem) for elem in self.invProb.phi_m_joint])
+        self.lambd.append("{:.2e}".format(self.invProb.lambd))
+        self.phi_c.append(self.invProb.phi_c)
+        self.phi.append(self.opt.f)
+
+        if self.save_txt:
+            f = open(self.fileName+'.txt', 'a')
+            f.write(
+                ' {0:2d}  {1}  {2}  {3}  {4}  {5:1.4e}  {6:d}  {7:1.4e}\n'.format(
+                self.opt.iter,
+                self.betas[self.opt.iter-1],
+                self.lambd[self.opt.iter-1],
+                self.phi_d[self.opt.iter-1],
+                self.phi_m[self.opt.iter-1],
+                self.phi_c[self.opt.iter-1],
+                self.opt.cg_count,
+                self.phi[self.opt.iter-1] )
+            )
+            f.close()
+
+    def load_results(self):
+        results = np.loadtxt(self.fileName+str(".txt"), comments="#")
+        self.betas = results[:, 1]
+        self.lambd = results[:, 2]
+        self.phi_d = results[:, 3]
+        self.phi_m = results[:, 4]
+        self.phi_c = results[:, 5]
+        self.f = results[:, 7]
+
+class BetaEstimate_ByEig_Joint(InversionDirective):
+
+    betas = None
+    beta0_ratio = 1
+
+    def initialize(self):
+        '''
+            Estimate initial betas by eigen values.
+            Method uses one iteration of the Power Method to estimate
+            eigenvectors for each separate inverse problem.
+        '''
+        m = self.invProb.model
+        f = self.invProb.getFields(m, store=True, deleteWarmstart=False)
+
+        # Fix the seed for random vector for consistent results
+        np.random.seed(0)
+        x0 = np.random.rand(*m.shape)
+
+        t = []
+        b = []
+        i_count = 0
+        # assume dmisfit and reg are combo objective functions.
+        # assume last regularization object is the coupling
+        for dmis, reg in zip(self.dmisfit.objfcts, self.reg.objfcts[:-1]):
```

Comment:
This is where the `utils.mat_utils.eigenvalue_by_power_iteration` will come into play.
something along the line (with optional renaming for clarity):
```
dmis_eigenvalues = []
reg_eigenvalues = []
 for dmis, reg in zip(self.dmisfit.objfcts, self.reg.objfcts[:-1]):
   dm_eigenvalues.append(eigenvalue_by_power_iteration(
            dmis, m, n_pw_iter=self.n_pw_iter,
        ))
        reg_eigenvalues.append(eigenvalue_by_power_iteration(
            reg, m, n_pw_iter=self.n_pw_iter,
        ))
```

Reply:
Updated